src/x509/verify.ts

```diff
@@ -0,0 +1,202 @@
+import { CertificateChainVerificationError } from '../error';
+import { x509Certificate } from './cert';
+
+interface VerifyCertificateChainOptions {
+  trustedCerts: x509Certificate[];
+  certs: x509Certificate[];
+  checkDate?: Date;
+}
+
+export function verifyCertificateChain(
+  opts: VerifyCertificateChainOptions
+): void {
+  const verifier = new CertificateChainVerifier(opts);
+  verifier.verify();
+}
+
+class CertificateChainVerifier {
+  private certs: x509Certificate[];
+  private trustedCerts: x509Certificate[];
+  private localCerts: x509Certificate[];
+  private checkDate: Date;
+
+  constructor(opts: VerifyCertificateChainOptions) {
+    this.certs = opts.certs;
+    this.trustedCerts = opts.trustedCerts;
+    this.localCerts = dedupeCertificates([...opts.trustedCerts, ...opts.certs]);
+    this.checkDate = opts.checkDate || new Date();
```

Comment:
It wasn't immediately obvious to me what `checkDate` refers to, wondering if something like `validAt` would be clearer?

Reply:
Renamed